### ISSUE TITLE
ratsubst error ZEROP: ((MMINUS) 1) is not a number

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#1361]
Imported from SourceForge on 2024-07-04 09:57:55
Created by **[billingd](https://sourceforge.net/u/billingd/)** on 2021-09-05 13:53:49
Original: https://sourceforge.net/p/maxima/bugs/3848

---

ratsubst gives an error ZEROP: ((MMINUS) 1) is not a number

This example is reduced from a larger expression.  I was unable to reduce it further.

~~~
(%i1) display2d:false$
(%i2)  d:-(3^(3/2)*%i+1)^(1/2)/3$
(%i3) ratsubst(y,5,d);
Maxima encountered a Lisp error:


ZEROP: ((MMINUS) 1) is not a number

Automatically continuing.
To enable the Lisp debugger set *debugger-hook* to nil.

(%i4) build_info();
(%o4) %build_info("5.45.0","2021-05-30 10:22:34","x86_64-pc-cygwin","CLISP",
                  "2.49+ (2010-07-17) (built on CygwinPro [192.168.48.1])",
                  "/home/dabilling/.maxima","/tmp",
                  "/home/dabilling/.maxima/binary/5_45_0/clisp/2_49___2010_07_17___built_on_CygwinPro__192_168_48_1__",
                  false,false)

~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:57:57
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2021-09-05 17:40:10
Original: https://sourceforge.net/p/maxima/bugs/3848/#6585

---

A smaller case that fails is: 
~~~
ratsubst(1,1,-1/2)
~~~

The problem is in ``nformat-all``. It is the same problem I mentioned in May (https://sourceforge.net/p/maxima/mailman/message/37288865/).

Using the "pragmatic" solution mentioned in that mail, here's the new code for ``nformat-all`` in **nforma.lisp**:

~~~
(defun nformat-all (form)
  (setq form (nformat form))
  (if (or (atom form)
	  (member (caar form) '(bigfloat rat))) ;can't recurse over cdrs
      form
    (cons (delsimp (car form))
	  (if (member (caar form) '(mdo mdoin))
	      (mapcar #'(lambda (u) (if u (nformat-all u))) (cdr form))
	    (mapcar #'nformat-all (cdr form))))))
~~~

The examples and the test suite pass with this change.

By the way, I noticed that ``delsimp`` in **comm.lisp** can be written more simply. Instead of
~~~
(defun delsimp (e)
  (delete 'simp (copy-list e) :count 1 :test #'eq))
~~~
just
~~~
(defun delsimp (e) (remove 'simp e))
~~~

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:01
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2021-09-06 16:11:44
Original: https://sourceforge.net/p/maxima/bugs/3848/#e24e

---

- **labels**:  --> ratsubst
- **Group**: None --> Includes_proposed_fix

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:04
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-02-17 17:05:07
Original: https://sourceforge.net/p/maxima/bugs/3848/#7564

---

- **status**: open --> pending

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:07
Created by **[macrakis](https://sourceforge.net/u/macrakis/)** on 2023-02-17 17:05:07
Original: https://sourceforge.net/p/maxima/bugs/3848/#27c3

---

Fixed in source.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:11
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-26 06:56:19
Original: https://sourceforge.net/p/maxima/bugs/3848/#ab6e

---

- **status**: pending --> open

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:14
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-02-26 06:56:19
Original: https://sourceforge.net/p/maxima/bugs/3848/#f6b1

---

Stavros, I've reopened this report since I don't see the bug fix in the source code. git blame says the last change to NFORMAT-ALL (in src/nforma.lisp) was a few years ago. Did the commit not get pushed or something like that?

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:18
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-03-09 05:22:03
Original: https://sourceforge.net/p/maxima/bugs/3848/#2985

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:21
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2023-03-09 05:22:04
Original: https://sourceforge.net/p/maxima/bugs/3848/#232a

---

I've replaced NFORMAT-ALL with the version shown here, and also DELSIMP. Fixed by commit 1e01557, closing this report.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 09:58:25
Created by **[billingd](https://sourceforge.net/u/billingd/)** on 2023-03-12 01:23:49
Original: https://sourceforge.net/p/maxima/bugs/3848/#c35e

---

New test in rtest6a.mac left variable d defined. Existing test in rtest6b.mac now fails.   Add kill(all) to end of rtest6a.mac to clean up.

Commit [36f3cc]
